sematic/plugins/storage/s3_storage.py

```diff
@@ -0,0 +1,169 @@
+# Standard Library
+import enum
+import logging
+from typing import Dict, Iterable, List, Type
+
+# Third-party
+import boto3
+import botocore.exceptions
+import requests
+
+# Sematic
+from sematic.abstract_plugin import (
+    AbstractPlugin,
+    AbstractPluginSettingsVar,
+    PluginVersion,
+)
+from sematic.config.settings import get_plugin_setting
+from sematic.plugins.abstract_storage import AbstractStorage, NoSuchStorageKey
+from sematic.utils.memoized_property import memoized_property
+from sematic.utils.retry import retry
+
+logger = logging.getLogger(__name__)
+
+_PLUGIN_VERSION = (0, 1, 0)
+
+
+class S3ClientMethod(enum.Enum):
+    PUT = "put_object"
+    GET = "get_object"
+
+
+class S3StorageSettingsVar(AbstractPluginSettingsVar):
+    AWS_S3_BUCKET_KEY = "AWS_S3_BUCKET"
+
+
+class S3Storage(AbstractStorage, AbstractPlugin):
+    """
+    Implementation of the `AbstractStorage` interface for AWS S3 storage. The
+    bucket where to store values is determined by the `AWS_S3_BUCKET` plug-in
+    setting.
+    """
+
+    PRESIGNED_URL_EXPIRATION = 5 * 60  # 5 minutes
+
+    @staticmethod
+    def get_author() -> str:
+        return "github.com/sematic-ai"
+
+    @staticmethod
+    def get_version() -> PluginVersion:
+        return _PLUGIN_VERSION
+
+    @classmethod
+    def get_settings_vars(cls) -> Type[AbstractPluginSettingsVar]:
+        return S3StorageSettingsVar
+
+    @memoized_property
+    def _bucket(self) -> str:
+        return get_plugin_setting(
+            self.__class__, S3StorageSettingsVar.AWS_S3_BUCKET_KEY
+        )
+
+    @memoized_property
+    def _s3_client(self):
+        return boto3.client("s3")
+
+    def _make_presigned_url(self, client_method: S3ClientMethod, key: str) -> str:
+        presigned_url = self._s3_client.generate_presigned_url(
+            ClientMethod=client_method.value,
+            Params={"Bucket": self._bucket, "Key": key},
+            ExpiresIn=self.PRESIGNED_URL_EXPIRATION,
+        )
+
+        return presigned_url
+
+    def _get_write_location(self, namespace: str, key: str) -> str:
+        return self._make_presigned_url(S3ClientMethod.PUT, f"{namespace}/{key}")
+
+    def _get_read_location(self, namespace: str, key: str) -> str:
+        return self._make_presigned_url(S3ClientMethod.GET, f"{namespace}/{key}")
+
+    def set(self, key: str, value: bytes):
+        """Store value in S3"""
+        logger.debug(f"{self.__class__.__name__} Setting value for key: {key}")
+
+        response = requests.put(key, data=value)
+        response.raise_for_status()
+
+    @retry(tries=3, delay=5)
+    def get(self, key: str) -> bytes:
+        """Get value from S3."""
+        response = requests.get(key)
```

Comment:
The `key` here seems to be a full URL. Is that the case? If so, is it the same `key` as in `get_line_stream()` (which seems to be a key in the bucket)?

Reply:
No it is not the same.

Here `get` implements the `AbstractStorage.get` interface for client-side code to access data. For cloud storage, pre-signed URLs are provided by the server. Therefore getting the data is only a matter of fetching the content of the URL.

`get_line_stream` below is an API specific to `S3Storage` with (currently) no equivalent on other storage engines (in the future, we should support logging on local machines, but we do not for now). `get_line_stream` is meant to be used server-side, which is aware of bucket info.

Note that these storage engines are used both client-side and server-side, where different permissions are available. I will add info to docstrings in `AbstractStorage`.

Reply:
Hmm, may I ask for some further clarifications?

> interface for client-side code to access data. For cloud storage, pre-signed URLs are provided by the server

Who is this `client-side` code? Is it the browser or something else? 

"_pre-signed URLs are provided by_ **the server**" who is the server? Do you refer to the S3 API or the Sematic server? 

Please bear with me since I am unclear about the background of the S3Storage. But I think the below points might be useful to spark some inspiration:

- Assuming a client (e.g. browser) already has a fully signed URL, why wouldn't it just fetch that URL by itself? Therefore the current implementation of `S3Storage::[AbstractStorage.get]()` doesn't do much to help.

- A common pattern I see in the FE-BE communication is that the BE returns a [302 redirect response](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/302) instead of reading the input stream and responding with an output stream. The client is intelligent enough to follow the `Location` header in the response and read that URL instead. Take the `AbstractStorage.get` (aka Read) as an example; it can be implemented in this way:

1. client requests `S3Storage/get/bucket_id`
1. server uses a secret to generate a temporary signed S3 URL 
1. server responds with 302 + Location: S3 URL (no byte stream data in the body at all)
1. client gets 302, automatically follows the S3 URL to read the bytestream.

Some advantages of this approach:
- Server doesn't transmit data. There will be no bottleneck choked on the server side. 
- Since there is less time processing on the server side and no data transmitted. We save $$$.
- Relying on the delivery network of a storage service like S3, its data *might be* globally distributed. So a client reading the signed URL by itself might be faster than reading from the server.
- From the client's point of view (say a browser), this API form `S3Storage/get/bucket_id` is more useful.

It's also worth mentioning that while the browser follows 302 redirections automatically. Many other clients do too. Python's `requests` library [has it](https://requests.readthedocs.io/en/latest/user/quickstart/?highlight=allow_redirects#redirection-and-history) turned on by default.

Reply:
My comment should not block this PR merge.

Reply:
Here "the client" is the resolver, not the browser.

Here is how it works:
1. User submits resolution, either local on remote.
2. Resolution proceeds and needs to persist artifacts.
3. The client (the resolver) does not know anything about where to store artifacts and has no storage permissions. Only the server knows where they should be stored (local machine or S3). The resolver queries the API for a location.
4. The server uses `AbstractStorage.get_location` with `mode=StorageMode.WRITE` to figure out where the artifacts should be stored
5. The plug-in infrastructure uses the correct storage engine based on the content of `settings.yaml`
6. The server returns to the client (the resolver) a location and a storage engine description.
7. The client (the resolver) uses the storage engine description to load up the correct storage engine and passes the location that was returned by the server to `AbstractStorage.set`.

In the case of S3, the location is a pre-signed URL returned to the resolver by the API.

In the case of a local API server, the location is a path on the user's machine.

For browser-side access to artifact data, first of all, it will be quite rare that the artifact data will be accessed. Usually, the content of the JSON summary that is in the DB is sufficient to display the artifact. I foresee some cases where artifact data will be used by the browser though. Notably: matplotlib figures since the JPEG has to be generated resolver-side, and Pandas dataframes if we ever want to display the entire dataframe in the UI.

When the browser will need to access the artifact data (e.g. matplotlib figure), the browser will use the same mechanism as the resolver:
1. the client (browser or resolver) will query the API server for a read location.
2. the server will use the relevant storage engine to generate a read location by calling `AbstractStorage.get_location` with `mode=StorageMode.READ`
3. the client (browser or resolver) will receive the location and query it directly

In the case of `S3Storage`, the read location will be a pre-signed URL (as can be seen in `S3Storage._get_read_location`.
In the case of `LocalStorage`, the read location will be an API endpoint that clients (resolver+browser) will query to get the bytes, see `LocalStorage._get_read_location`.
The resolver is able to access the filesystem, but the browser is not, therefore we need to go through the API server to read local data.
That is slightly inefficient for the resolver (reading bytes through the API server instead of directly from disk), but this is immaterial since everything is running locally.

Reply:
Thanks @neutralino1 for this detailed explanation.

> the client (browser or resolver) will query the API server for a read location.
the server will use the relevant storage engine to generate a read location by calling AbstractStorage.get_location with mode=StorageMode.READ
the client (browser or resolver) will receive the location and query it directly

This is the use case I propose to simplify. If we adopt 302 redirections, this becomes a single API point, regardless of a local file or S3. 
- Assume this API is called `GET api/artifact/[id]`
- Then in S3 cases this become `304, Location: S3_signed_read_URL`
- in case of local file, this becomes ```response.write(open(file_path, "r"))``` (And unfortunately, in this case, the server does transmit data)

So the concept of storage location is hidden from the client's perspective. The client directly gets the byte stream.

Separately, from your description, it seems that in either Resolver or Browser cases, they expect to get the location URL or Path.

> In the case of S3, the location is a pre-signed URL returned to the resolver by the API.

> In the case of a local API server, the location is a path on the user's machine.

So why do we still need this method in `s3_storage.py` ?
```
@retry(tries=3, delay=5)
    def get(self, key: str) -> bytes:
        """Get value from S3."""
        response = requests.get(key)
```
why wouldn't the client directly request the URL?

Lastly, one use case for Browser to directly read data might be reading logs directly from S3.

Reply:
Actually there was a slight incorrect statement in my message.
For local storage, the resolver reads the bytes directly from disk based on the location returned by the server, as can be seen in `LocalStorage.get`.

In this case, redirection would not work because it would mean that in one case the server would redirect to the content (S3 storage), in another, the server would provide a local path for the resolver to read from. This introduces a few layers of indirection. Also, the server endpoint is not aware of what to do (redirect vs return bytes). The endpoint simply uses the storage engine and gets a read location. If we wanted to implement the redirect-or-read option, we would have to make the endpoint do a conditional branching based on `LocalStorage` vs` S3Storage`, which breaks the concept of plug-ins (i.e. anyone can implement a storage engine as long as the interface is respected).

> So why do we still need this method in `s3_storage.py` ?

Same reason, it is a matter of unifying the interface across storage engines. What to do with the location returned by the server depends on the storage engine (e.g. read local file vs. query signed URL), that is what `.get` implements.